Products/ZenHub/metricpublisher/tests/test_publisher.py

```diff
@@ -0,0 +1,441 @@
+##############################################################################
+#
+# Copyright (C) Zenoss, Inc. 2014, all rights reserved.
+#
+# This content is made available according to terms specified in
+# License.zenoss under the directory where your Zenoss product is installed.
+#
+##############################################################################
+
+from unittest import TestCase, TestSuite, makeSuite
+from mock import Mock, MagicMock, create_autospec, patch
+from zope.interface.verify import verifyObject
+
+import ujson as json
+
+from Products.ZenHub.metricpublisher.publisher import (
+    os,
+    sys,
+    defer,
+    deque,
+    RedisListPublisher,
+    HttpPostPublisher,
+    BasePublisher,
+    StringProducer,
+    ResponseReceiver,
+    IBodyProducer,
+    RedisClientFactory,
+    Headers,
+    CookieJar,
+    CookieAgent,
+    UNAUTHORIZED,
+    INITIAL_REDIS_BATCH,
+    defaultPublishFrequency,
+    defaultMetricBufferSize,
+    basic_auth_string_content
+)
+
+
+METRIC_BUILDED = [
+    '{"timestamp":1535460634.26479,"metric":"eventQueueLength",\
+    "value":0.0,"tags":{\
+        "instance":"0","daemon":"zenpython","monitor":"localhost",\
+        "metricType":"GAUGE","tenantId":"b0e4t72hrole5z1xv88djc1hb"\
+        }\
+    }'
+]
+METRIC = 'testMetricName'
+BUFFER_LEN = 10
+PUBLISHER_FREQ = 10
+SCHEDULED = True
+
+
+class PublishError(Exception):
```

Comment:
Any time we need to define a class, function, or object just for testing, it is a good candidate for a Mock.

for example, we can replace this class with
```
PublishError = Mock(Exception, name='PublishError', autospec=True)
```

This object is only used in a single test, `BasePublisherTest.test_publish_failed` and should be moved into that test.